feat: implement MMU & MMIO

powerslider opened the pull request.

powerslider committed:
feat(mmio): add generated trace

powerslider committed:
fix: MMU and MMIO constants generation according to new tracing format

delehef commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmu/MicroData.java:
```diff
@@ -129,22 +129,22 @@ class MicroData {
+            Trace.PushTwoStackToRam,
+            Trace.StoreXInAThreeRequired,
+            Trace.StoreXInB,
+            Trace.StoreXInC)
```
Please put that in a static member, and a set

powerslider committed:
feat(mmio): add current trace and ram to ram dispatcher

powerslider committed:
feat(mmio): add RamToRamDispatcher, RamToRamSlideChunkDispatcher, RamToRamSlideOverlappingChunkDispatcher

powerslider committed:
feat(mmio): add RamLimbExcisionDispatcher

powerslider committed:
feat(mmio): add NaRamToStack3To2FullDispatcher

powerslider committed:
feat(mmio): add PushTwoRamToStackFullDispatcher

powerslider committed:
feat(mmio): add NaRamToStack1To1PaddedAndZeroDispatcher, PushOneRamToStackDispatcher

powerslider committed:
feat(mmio): add NaRamToStack2To1PaddedAndZeroDispatcher

powerslider committed:
feat(mmio): add FirstFastSecondPaddedDispatcher, FirstPaddedtSecondZeroDispatcher, NaRamToStack2To2PaddedDispatcher, NaRamToStack3To2PaddedDispatcher

powerslider committed:
feat(mmio): move all dispatchers to dispatchers package

powerslider committed:
feat(mmio): add ExceptionalRamToStack3To2FullDispatcher, ExceptionalRamToStack3To2FullFastDispatcher, NaRamToStack2To1FullAndZeroDispatcher

powerslider committed:
feat(mmio): add FullStackToRamDispatcher, LsbFromStackToRamDispatcher, PushTwoStackToRamDispatcher

powerslider committed:
feat(mmio): modify MmioDispatcher interface and group all dispatcher implementations

powerslider committed:
feat(mmio): add all missing dispatcher implementations

powerslider committed:
fix(mmio): small fixes

powerslider committed:
feat(mmio): add last dispatcher implementations

powerslider committed:
feat(mmio): hook up all dispatchers

powerslider committed:
fix: extract MicroData and related classes to runtime/microdata package and connect MMU to HUB and MMU to MMIO (still more work needed)

powerslider committed:
fix: connect MMU to MMIO and correctly instantiate both in the HUB

powerslider committed:
fix: connect MMU to MMIO and correctly instantiate both in the HUB

powerslider committed:
fix: some MMU fixes

powerslider committed:
fix: line count prediction mismatch issue in MMU

powerslider committed:
fix: adapt MMU and MMIO to the new cached line count mechanism

delehef commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmio/MmioOperation.java:
```diff
@@ -15,12 +15,24 @@
-    boolean isInitCode) {}
+@RequiredArgsConstructor
+@Accessors(fluent = true)
+class MmioOperation extends ModuleOperation {
```
If all attributes are `@Getter`-annotated, just annotate the class itself with `@Getter`

delehef commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmu/MicroDataProcessor.java:
```diff
@@ -70,6 +70,10 @@ MicroData dispatchOpCode(
     return currentPreComputation.dispatch(microData, opCode, stackOps, callStack);
   }
 
+  static int maxCounter(final boolean oob) {
```
Should this be an instance method instead of taking an attribute as argument?

powerslider committed:
fix: MMIO fixes

powerslider committed:
feat: re-generate MMU and MMIO traces

powerslider committed:
feat: re-generate MMU and MMIO traces

powerslider committed:
fix: replace manual copy with system.arraycopy for FirstFastSecondPaddedDispatcher

powerslider committed:
fix: call MMIO handleRam in MMU handleRam and outside of MMU commit method

powerslider committed:
fix: extract fast micro instructions into a set constrant

powerslider committed:
fix: fix all failing tests

powerslider committed:
fix: unify format of maxCounter method for MMU and MMIO

powerslider committed:
fix: rebase with main

delehef commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmu/Mmu.java:
```diff
@@ -129,53 +138,54 @@ private void traceMicroData(MicroData microData, final CallStack callStack, Trac
   private void trace(MicroData microData, Trace trace) {
```
Should be a method of `MicroData`

powerslider replied:
Why not `MmuOperation`?

powerslider commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmio/MmioOperation.java:
```diff
@@ -0,0 +1,41 @@
+
+  @Override
+  protected int computeLineCount() {
+    return 1 + maxCounter(microData);
```
`MicroData` was extracted to be outside of MMU and MMIO modules, because it is equally shared among the and I wanted to avoid duplex dependency between them. `maxCounter` method has specific implementations tightly related to MMU and MMIO and just depend on some data from `MicroData`, so it did not make sense to me to include it there, because it was module specific.

delehef replied:
Makes sense, agreed.

powerslider commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/mmio/Mmio.java:
```diff
@@ -15,30 +15,154 @@
+        new MmioOperation(microData, mmioDataProcessor, moduleStamps, microStamp, isInitCode));
+  }
+
+  private void trace(
```
Will do.

powerslider commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/memory/Memory.java:
```diff
@@ -0,0 +1,71 @@
+
+import net.consensys.linea.zktracer.types.UnsignedByte;
+
+public class Memory {
```
What would you suggest for this one?

powerslider replied:
I ported the name from zk-geth.

delehef replied:
Let's say e.g. `MemorySegmentSnapshot`

powerslider committed:
fix: some of the PR remarks

powerslider commented on arithmetization/src/test/java/net/consensys/linea/zktracer/testing/BytecodeRunner.java:
```diff
@@ -73,7 +73,7 @@ public void run() {
             .to(receiverAccount)
             .keyPair(keyPair)
-            .gasLimit(100_000_000L)
+            .gasLimit(50_000_000_000L)
```
This was modified due to MXP test failures, as far as I can remember. We decided that with @lorenzogentile404 and it fixed them after we increased the gas limit.

powerslider commented on arithmetization/src/main/java/net/consensys/linea/zktracer/runtime/microdata/MicroData.java:
```diff
@@ -13,35 +13,60 @@
+  private static final int PUSH_TWO_STACK_TO_RAM = 608;
+  private static final int STORE_X_IN_A_THREE_REQUIRED = 609;
+  private static final int STORE_X_IN_B = 610;
+  private static final int STORE_X_IN_C = 611;
```
NOTE: We agreed on simply generating a new trace file that only contains the constants and add a `microdata` task in `trace-files.gradle`

powerslider committed:
fix: add more PR remarks

powerslider committed:
fix: fix build issues

powerslider committed:
fix: fix lots of issues related to NPEs, wrong source and context IDs

powerslider committed:
feat: generate microdata fast microinstructions constants based on corset constraints

powerslider committed:
feat: generate microdata fast microinstructions constants based on mmu corset constants definitions

delehef commented on arithmetization/src/main/java/net/consensys/linea/zktracer/container/stacked/set/StackedSet.java:
```diff
@@ -121,7 +122,8 @@ public <T> T[] toArray(@NotNull T[] a) {
   @Override
   public boolean add(E e) {
-    final boolean isNew = this.sets.peekLast().add(e);
+    final boolean isNew = Optional.ofNullable(this.sets.peekLast()).orElse(new HashSet<>()).add(e);
```
We don't want this: there is no correct use case of the `StackedSet` where we never called `enter()`.

Either revert it or throw an exception, but don't sneakily insert a new set in it.

powerslider replied:
Ok, I will revert it.

powerslider committed:
feat(mmu-v2): new trace

letypequividelespoubelles committed:
feat: preparing tracing

letypequividelespoubelles committed:
fix(mmu): columns forgotten

letypequividelespoubelles committed:
feat: ras

letypequividelespoubelles committed:
feat(txnData): add check callData is zero (#580)

* feat(txnData): add check callData is zero

* style: spotless

* style: column renaming

* fix(constraint): ras

* feat(constraint): ras

letypequividelespoubelles committed:
feat(txnData): add check callData is zero (#580)

* feat(txnData): add check callData is zero

* style: spotless

* style: column renaming

* fix(constraint): ras

* feat(constraint): ras

letypequividelespoubelles committed:
perf(BIN): adaptative number line (#531)

* perf(BIN): adaptative nb line

* fix(bin): rebase fix

* feat: column typing

* fix(constraint): ras

* fix: ras

* fix: corset export

* feat(constraint): rebase

* feat(constraint): ras

letypequividelespoubelles committed:
perf(BIN): adaptative number line (#531)

* perf(BIN): adaptative nb line

* fix(bin): rebase fix

* feat: column typing

* fix(constraint): ras

* fix: ras

* fix: corset export

* feat(constraint): rebase

* feat(constraint): ras

letypequividelespoubelles committed:
style: ras + cleaning

letypequividelespoubelles committed:
feat(rlp*): add phase column (#586)

* feat(constraint): ras

* build: fix RLP modules trace creation

* feat: fradle Trace

* feat(rlp*): add phase column

* style(rlpTxn): phase starts at 1 instead of 0

* feat(constraint): ras

---------

Co-authored-by: Franklin Delehelle <franklin.delehelle@odena.eu>

letypequividelespoubelles committed:
feat(rlp*): add phase column (#586)

* feat(constraint): ras

* build: fix RLP modules trace creation

* feat: fradle Trace

* feat(rlp*): add phase column

* style(rlpTxn): phase starts at 1 instead of 0

* feat(constraint): ras

---------

Co-authored-by: Franklin Delehelle <franklin.delehelle@odena.eu>

Commit:
feat[hub]: implement the precompile call mechanism & plug the OOB (#581)

Signed-off-by: franklin.delehelle@consensys.net
Co-authored-by: Fabio Di Fabio <fabio.difabio@consensys.net>

Commit:
feat[hub]: implement the precompile call mechanism & plug the OOB (#581)

Signed-off-by: franklin.delehelle@consensys.net
Co-authored-by: Fabio Di Fabio <fabio.difabio@consensys.net>

powerslider committed:
feat(blake2f-modexp-data): transform MODEXP_DATA to BLAKE2f_MODEXP_DATA (#576)

Resolves: #575
Signed-off-by: Tsvetan Dimitrov tsvetan.dimitrov@consensys.net

powerslider committed:
feat(blake2f-modexp-data): transform MODEXP_DATA to BLAKE2f_MODEXP_DATA (#576)

Resolves: #575
Signed-off-by: Tsvetan Dimitrov tsvetan.dimitrov@consensys.net

powerslider committed:
fix(blake2f-modexp-data) missed fixes (#608)

*  Add @EqualsAndHashCode.Include to hubStamp.
*  Calculate line count based on which precompile is called.

Signed-off-by: Tsvetan Dimitrov <tsvetan.dimitrov@consensys.net>

powerslider committed:
fix(blake2f-modexp-data) missed fixes (#608)

*  Add @EqualsAndHashCode.Include to hubStamp.
*  Calculate line count based on which precompile is called.

Signed-off-by: Tsvetan Dimitrov <tsvetan.dimitrov@consensys.net>

letypequividelespoubelles committed:
fix(shf,bin): lookup to reference table (#613)

letypequividelespoubelles committed:
fix(shf,bin): lookup to reference table (#613)

lorenzogentile404 committed:
feat: implement the EXP module

Fixes #303 

---------

Co-authored-by: Tsvetan Dimitrov <tsvetan.dimitrov@consensys.net>
Co-authored-by: Franklin Delehelle <franklin.delehelle@odena.eu>
Co-authored-by: delehef <franklin.delehelle@consensys.net>

Commit:
perfs: limit the count of post-tracing loops (#609)

Signed-off-by franklin.delehelle@consensys.net

letypequividelespoubelles committed:
fix: debug

Commit:
fix: do not prune failing transactions from the replays

Dropping failing transactions initial state is plain stupid, as they will fail due to their initial state.

letypequividelespoubelles committed:
fix(romLex): wrong CFI implementation (#633)



Co-authored-by: Franklin Delehelle <franklin.delehelle@odena.eu>

Commit:
fix: debug Hub arithmetization – first wave (#625)


Signed-off-by: franklin.delehelle@consensys.net
Co-authored-by: Fabio Di Fabio <fabio.difabio@consensys.net>

Commit:
fix: gather & re-throw tracing exceptions (#637)

Signed-off-by: franklin.delehelle@consensys.net

delehef committed:
fix: do not overflow when reading logs payload

letypequividelespoubelles committed:
feat: fix total size must be given for anyToRam and compute mmuInstFlag only once

powerslider committed:
fix: rebase to arith-dev and fixup some part of the history

powerslider committed:
fix(mmio-v2): add missing docs to ExoSumDecoder

powerslider committed:
fix(mmio-v2): add missing docs to ExoToRamTransplants

powerslider committed:
fix(mmio-v2): add missing docs to TxTrace

powerslider committed:
Merge branch 'arith-dev' into feat/issue-142/mmio

powerslider committed:
fix(mmio-v2): missing fixes

letypequividelespoubelles committed:
fix(constraints): ras

letypequividelespoubelles committed:
feat(constraints): update

powerslider committed:
Merge branch 'arith-dev' into feat/issue-142/mmio

letypequividelespoubelles committed:
feat: generate trace for rlpTxn and txnData

letypequividelespoubelles committed:
feat(constraints): cleaning

letypequividelespoubelles committed:
fix: add todo

letypequividelespoubelles committed:
fix: some target ram limb might be touched by two consecutive mmio instructions

letypequividelespoubelles committed:
fix(mmu): fill the LIMB column when dealing with exo mmio inst

letypequividelespoubelles committed:
feat(blake): get blake modexp data from the mmu

letypequividelespoubelles committed:
fix: make it compile

letypequividelespoubelles committed:
fix(rlpTxn): typo

powerslider committed:
fix: attach calling the MMU from the ImcFragment and perform a check if the current tx requires EVM execution

letypequividelespoubelles committed:
fix: mload, mstore, mstore8 debug

powerslider committed:
fix: registered missing mmio instructions

letypequividelespoubelles committed:
fix: debug mstore and mload

letypequividelespoubelles committed:
feat(constraints): update

letypequividelespoubelles committed:
fix(mmio): right column name

letypequividelespoubelles committed:
test(rlpTxn): commenting random rlpTxnTest

letypequividelespoubelles committed:
test: add txInit and Create test

letypequividelespoubelles committed:
fix(mmu): confusion between dividend and divisor

letypequividelespoubelles committed:
fix: callStackReader

letypequividelespoubelles committed:
fix: move to commit time filling exoLimb

letypequividelespoubelles committed:
fix(mmio-pattern): typo

letypequividelespoubelles committed:
fix: mmu inst invalid code prefix must access the source ram bytes during the preprocessing

letypequividelespoubelles committed:
fix: snapshot of the ram memory at pre opcode time

letypequividelespoubelles committed:
test: add overlapping mstore mload test

letypequividelespoubelles committed:
fix: regenerate Trace.java for the hub as it was using wrong constants

letypequividelespoubelles committed:
test: merge mload mstore test into one more complex test

letypequividelespoubelles committed:
fix: typo + add TODO, missing all exoSum + other issue in the hub triggering

letypequividelespoubelles committed:
fix(hub): typo

letypequividelespoubelles committed:
style(rlpTxn): column renaming

letypequividelespoubelles committed:
feat(constraints): update

letypequividelespoubelles committed:
fix: typo

letypequividelespoubelles committed:
fix(mmu): fixing exoSum and empty ram for fictious cn with copied callData

powerslider committed:
fix: patch CallStackReader.valueFromMemory to handle MANTLE call frame

letypequividelespoubelles committed:
fix(mmu): debug right padded word extraction

letypequividelespoubelles committed:
fix: simplify how the hub gives the exoSum to the mmu

letypequividelespoubelles committed:
fix(mmu): missing size in inst needed for lookup to rlpTxn

letypequividelespoubelles committed:
fix: debugged ramToExoWithPadding

letypequividelespoubelles committed:
fix: empty mmuCall and fix BlakeModexp

letypequividelespoubelles committed:
fix: euc shouldn't trace doublon operation

letypequividelespoubelles committed:
fix: put long instead of int where needed

letypequividelespoubelles committed:
fix(constraints): update

letypequividelespoubelles committed:
fix: euc byte decomposition

letypequividelespoubelles committed:
Merge branch 'feat/issue-142/mmio' of github.com:Consensys/linea-besu-plugin into feat/issue-142/mmio

letypequividelespoubelles committed:
fix(hub): callData information and depth of root context is 0, not 1

letypequividelespoubelles committed:
fix(constraints): small update

letypequividelespoubelles committed:
fix: regenerate Trace.java

letypequividelespoubelles committed:
fix(hub): try to not trigger the MMU when exceptions, but seems it doesn't work

letypequividelespoubelles committed:
fix: no exoFlag when not needed

letypequividelespoubelles committed:
fix(hub): fix hub triggering mmu for zero length

letypequividelespoubelles committed:
fix: hub triggerig of the mmu

letypequividelespoubelles committed:
fix: lookup mmio to rlpTxn too much constrained

letypequividelespoubelles committed:
fix(hub): at txInit, callFrame at h+1 is NOT a Mantle, only the one at h

letypequividelespoubelles committed:
style: ras, spotless

letypequividelespoubelles committed:
docs: add required TODO for sha/rip precompile

letypequividelespoubelles committed:
feat: add a new type of callFrame for precompile results

letypequividelespoubelles committed:
fix: fix hub calling mmu for return and revert

letypequividelespoubelles committed:
style: spotless + constraint submodule update

powerslider committed:
feat: merge with changes related to sequencer code removal

letypequividelespoubelles committed:
feat: add fictitious callFrame for precompileResult

letypequividelespoubelles committed:
fix: debug some more mmu Inst

letypequividelespoubelles committed:
Merge branch 'feat/issue-142/mmio' of github.com:Consensys/linea-arithmetization into feat/issue-142/mmio

letypequividelespoubelles committed:
style: spotless

letypequividelespoubelles committed:
fix: debug almost finish ?

letypequividelespoubelles committed:
fix: small and ras fixes

letypequividelespoubelles committed:
fix(hub): the mmu triggering of precompile is differed, so must take a snap of the hubStamp BEFORE defering

letypequividelespoubelles committed:
feat(constraints): uodtae
